sql/src/main/java/org/apache/druid/sql/calcite/schema/DruidSchema.java

```diff
@@ -270,8 +270,15 @@ public void start() throws InterruptedException
                   // Add missing segments back to the refresh list.
                   segmentsNeedingRefresh.addAll(Sets.difference(segmentsToRefresh, refreshed));
 
+                  // remove any tables which have been completely dropped before refresh
```

Comment:
This looks like it's undoing a mistake in removeSegment — the datasource should never have been added to dataSourcesNeedingRebuild in the first place. IMO it'd be better to make the change there. Doing it here works, though, so I'm OK with it if you don't want to move the logic around.

Reply:
Ah, the removal of segments from the broker could also mean that a datasource is no longer global/joinable and so also could require a rebuild, so i'm not sure we can remove it anymore without rethinking/reworking some other stuff. Since there were multiple places that we add to `dataSourcesNeedingRebuild`, but only one place we add to `tables`, it seemed less room for error to do the filtering here in the only place we populate it.

Reply:
`removeSegment` is only called when the last replica of a segment is gone, though.

Reply:
Oh good point, the broker thing happens in `removeServerSegment`, not `removeSegment`. Ok, i'll remove the `dataSourcesNeedingRebuild.add` from `removeSegment` :+1:

Reply:
I think it should be there, but only called if there's some other segments still remaining.

Reply:
Looks like one of the integration tests failed after the latest modifications, so more than just moving `dataSourcesNeedingRebuild.add` to the else is needed. I suspect it is the `removeServerSegment` which should also probably check if there are any segments before adding to the list in case it is triggered by the broker (broker segments are not currently tracked in `segmentMetadataInfo` since they will always currently be duplicated on a historical)